hbase-server/src/main/java/org/apache/hadoop/hbase/master/MobFileCompactionChore.java

```diff
@@ -0,0 +1,179 @@
+          if (hcd.isMobEnabled()) {
+            if (!reported) {
+              master.reportMobCompactionStart(htd.getTableName());
+              reported = true;
+            }
+            LOG.info(" Major compacting "+ htd.getTableName() + " cf=" + hcd.getNameAsString());
+            if (regionBatchSize == MobConstants.DEFAULT_MOB_MAJOR_COMPACTION_REGION_BATCH_SIZE) {
+              admin.majorCompact(htd.getTableName(), hcd.getName());
+            } else {
+              performMajorCompactionInBatches(admin, htd, hcd);
+            }
+          }
+        }
+        if (reported) {
+          master.reportMobCompactionEnd(htd.getTableName());
+          reported = false;
+        }
+      }
+    } catch (Exception e) {
+      LOG.error("Failed to compact", e);
+    } finally {
+      running = false;
+    }
+  }
+
+  private void performMajorCompactionInBatches(Admin admin, TableDescriptor htd,
+      ColumnFamilyDescriptor hcd) throws IOException {
+
+    List<RegionInfo> regions = admin.getRegions(htd.getTableName());
+    if (regions.size() <= this.regionBatchSize) {
+      admin.majorCompact(htd.getTableName(), hcd.getName());
+      return;
+    }
+    // Shuffle list of regions in case if they come ordered by region server
+    Collections.shuffle(regions);
+    // Create first batch
+    List<RegionInfo> toCompact = new ArrayList<RegionInfo>();
+    for (int i=0; i < this.regionBatchSize; i++) {
+      toCompact.add(regions.remove(0));
+    }
+
+    // Start compaction now
+    for(RegionInfo ri: toCompact) {
+      startCompaction(admin, htd.getTableName(), ri);
+    }
+
+    List<RegionInfo> compacted = new ArrayList<RegionInfo>();
+    while(!toCompact.isEmpty()) {
+      // Check status of active compactions
+      for (RegionInfo ri: toCompact) {
+        if (admin.getCompactionStateForRegion(ri.getRegionName()) == CompactionState.NONE) {
+          LOG.info("Finished major compaction: table={} region={}", htd.getTableName(),
+            ri.getRegionNameAsString());
+          compacted.add(ri);
+        }
+      }
+      // Update batch: remove compacted regions and add new ones
+      for (RegionInfo ri: compacted) {
+        toCompact.remove(ri);
+        if (regions.size() > 0) {
+          RegionInfo region = regions.remove(0);
+          startCompaction(admin, htd.getTableName(),region);
+          toCompact.add(region);
+        }
+      }
+      compacted.clear();
+      try {
+        Thread.sleep(1000);
+      } catch(InterruptedException e) {
+        // swallow
+      }
+    }
+    LOG.info(" Finished major compacting "+ htd.getTableName() + " cf=" + hcd.getNameAsString());
+
+  }
+
+  private void startCompaction(Admin admin, TableName table,  RegionInfo region)
+      throws IOException {
+
+    LOG.info("Started major compaction: table={} region={}", table,
+      region.getRegionNameAsString());
+    admin.majorCompactRegion(region.getRegionName());
```

Comment:
should be the version of `majorCompactRegion` that includes a column family to avoid duplicating work when a table has multiple column families that are mob enabled.

Reply:
Done.